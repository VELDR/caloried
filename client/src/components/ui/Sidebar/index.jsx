import PropTypes from 'prop-types';
import { useState } from 'react';
import { injectIntl } from 'react-intl';
import { useLocation, useNavigate } from 'react-router-dom';
import { Tooltip } from '@mui/material';
import { AutoStories, ChevronLeft, ChevronRight, Dashboard, Fastfood, Person } from '@mui/icons-material';

import classes from './style.module.scss';

const Sidebar = ({ intl: { formatMessage } }) => {
  const [isOpen, setIsOpen] = useState(false);
  const navigate = useNavigate();
  const location = useLocation();

  const toggleSidebar = () => {
    setIsOpen(!isOpen);
  };

  const isActive = (path) => location.pathname === path;

  const getItemClassName = (path) => `${classes.sidebar__item} ${isActive(path) ? classes.active : ''}`;

  return (
    <div className={classes.container} data-testid="sidebar">
      {isOpen && (
        <div className={classes.sidebar}>
          <div
            className={getItemClassName('/dashboard')}
            onClick={() => navigate('/dashboard')}
            data-testid="side-navigate-dashboard"
          >
            <Tooltip arrow title={formatMessage({ id: 'app_dashboard' })}>
              <Dashboard />
            </Tooltip>
          </div>
          <div
            className={getItemClassName('/diary')}
            onClick={() => navigate('/diary')}
            data-testid="side-navigate-diary"
          >
            <Tooltip arrow title={formatMessage({ id: 'app_diary' })}>
              <AutoStories />
            </Tooltip>
          </div>
          <div
            className={getItemClassName('/search')}
            onClick={() => navigate('/search')}
            data-testid="side-navigate-search"
          >
            <Tooltip arrow title={formatMessage({ id: 'app_food' })}>
              <Fastfood />
            </Tooltip>
          </div>
          <div
            className={getItemClassName('/profile')}
            onClick={() => navigate('/profile')}
            data-testid="side-navigate-profile"
          >
            <Tooltip arrow title={formatMessage({ id: 'app_profile' })}>
              <Person />
            </Tooltip>
          </div>
        </div>
      )}

      <div className={classes.chevron} onClick={toggleSidebar} data-testid="toggle-sidebar">
        {isOpen ? <ChevronLeft /> : <ChevronRight />}
      </div>
    </div>
  );
};

Sidebar.propTypes = {
  intl: PropTypes.object,
};

export default injectIntl(Sidebar);
