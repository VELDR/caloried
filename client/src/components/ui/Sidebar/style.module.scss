@import '@styles/base';

.container {
  @include flex-center;
  position: fixed;
  left: 0;
  top: 50%;
  transform: translateY(-50%);
  z-index: 100;
  display: none;

  @include apply-at-sm-size {
    display: flex;
  }
  .sidebar {
    @include flex-center-column(2rem);
    width: fit-content;
    transition: width 0.3s ease;
    background-color: var(--color-bg-primary);
    border: 2px solid var(--color-text-primary);
    border-top-right-radius: 1rem;
    border-bottom-right-radius: 1rem;
    border-left: none;

    &__item {
      @include flex-column;
      align-items: center;
      font-size: $font-size-small;
      color: var(--color-text-primary);

      overflow: hidden;
      cursor: pointer;

      @include apply-at-md-size {
        @include flex-items-center(0.25rem);
        flex-direction: row;
        width: 100%;
        transition:
          width 200ms,
          padding 200ms;
        transition-timing-function: ease-in-out;
      }
      &:hover {
        svg {
          font-size: $font-size-huge;
        }
      }
      svg {
        font-size: $font-size-larger;
        position: relative;
        z-index: 2;
        transition:
          font-size 250ms,
          padding 250ms;
        transition-timing-function: ease-out;
        padding: 0.5rem;
      }
    }
    .active {
      svg {
        color: var(--color-primary);
      }
    }
  }
  .chevron {
    cursor: pointer;
    @include flex-center;

    padding: 0.5rem 0;
    border-top-right-radius: 0.5rem;
    border-bottom-right-radius: 0.5rem;
    color: var(--color-bg-primary);
    background-color: var(--color-text-primary);
    svg {
      font-size: $font-size-medium;
    }
    @include apply-at-xl-size {
      padding: 0.5rem 0;
      svg {
        font-size: $font-size-large;
      }
    }
    &:hover {
      color: var(--color-text-primary);
      background-color: var(--color-bg-primary);
      border: 1px solid var(--color-text-primary);
      border-left: none;
    }
  }
}
