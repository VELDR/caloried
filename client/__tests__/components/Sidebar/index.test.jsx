import Sidebar from '@components/ui/Sidebar';
import { fireEvent, render } from '@utils/testHelper';

const mockNavigate = jest.fn();
const mockLocation = {
  pathname: '/dashboard',
};

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
  useLocation: () => mockLocation,
}));

let wrapper;

beforeEach(() => {
  wrapper = render(<Sidebar />);
  const { getByTestId } = wrapper;
  const toggleButton = getByTestId('toggle-sidebar');
  fireEvent.click(toggleButton);
});

describe('Sidebar Component', () => {
  it('should render correctly', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('sidebar')).toBeInTheDocument();
  });
  it('should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should call navigate to dashboard when dashboard button is clicked', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('side-navigate-dashboard');
    fireEvent.click(button);
    expect(mockNavigate).toHaveBeenCalledWith(`/dashboard`);
  });
  it('should call navigate to diary when diary button is clicked', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('side-navigate-diary');
    fireEvent.click(button);
    expect(mockNavigate).toHaveBeenCalledWith(`/diary`);
  });
  it('should call navigate to search when food button is clicked', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('side-navigate-search');
    fireEvent.click(button);
    expect(mockNavigate).toHaveBeenCalledWith(`/search`);
  });
  it('should call navigate to profile when profile button is clicked', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('side-navigate-profile');
    fireEvent.click(button);
    expect(mockNavigate).toHaveBeenCalledWith(`/profile`);
  });
  it('should apply the correct classes for active and inactive buttons', () => {
    const { getByTestId } = wrapper;

    const dashboardButton = getByTestId('side-navigate-dashboard');
    expect(dashboardButton).toHaveClass('active');
    expect(dashboardButton).toHaveClass('sidebar__item');

    const diaryButton = getByTestId('side-navigate-diary');
    expect(diaryButton).not.toHaveClass('active');
    expect(diaryButton).toHaveClass('sidebar__item');

    mockLocation.pathname = '/diary';
    wrapper.rerender(<Sidebar />);

    expect(diaryButton).toHaveClass('active');
  });
});
