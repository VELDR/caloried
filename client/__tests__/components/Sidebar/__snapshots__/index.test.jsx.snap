// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Component should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
        data-testid="sidebar"
      >
        <div
          class="sidebar"
        >
          <div
            class="sidebar__item active"
            data-testid="side-navigate-dashboard"
          >
            <svg
              aria-hidden="true"
              aria-label="Dashbor"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="DashboardIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 13h8V3H3v10zm0 8h8v-6H3v6zm10 0h8V11h-8v10zm0-18v6h8V3h-8z"
              />
            </svg>
          </div>
          <div
            class="sidebar__item "
            data-testid="side-navigate-diary"
          >
            <svg
              aria-hidden="true"
              aria-label="Diary"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="AutoStoriesIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m19 1-5 5v11l5-4.5V1zM1 6v14.65c0 .25.25.5.5.5.1 0 .15-.05.25-.05C3.1 20.45 5.05 20 6.5 20c1.95 0 4.05.4 5.5 1.5V6c-1.45-1.1-3.55-1.5-5.5-1.5S2.45 4.9 1 6zm22 13.5V6c-.6-.45-1.25-.75-2-1v13.5c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5v2c1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.3 4.75 1.05.1.05.15.05.25.05.25 0 .5-.25.5-.5v-1.1z"
              />
            </svg>
          </div>
          <div
            class="sidebar__item "
            data-testid="side-navigate-search"
          >
            <svg
              aria-hidden="true"
              aria-label="Makanan"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="FastfoodIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18.06 22.99h1.66c.84 0 1.53-.64 1.63-1.46L23 5.05h-5V1h-1.97v4.05h-4.97l.3 2.34c1.71.47 3.31 1.32 4.27 2.26 1.44 1.42 2.43 2.89 2.43 5.29v8.05zM1 21.99V21h15.03v.99c0 .55-.45 1-1.01 1H2.01c-.56 0-1.01-.45-1.01-1zm15.03-7c0-8-15.03-8-15.03 0h15.03zM1.02 17h15v2h-15z"
              />
            </svg>
          </div>
          <div
            class="sidebar__item "
            data-testid="side-navigate-profile"
          >
            <svg
              aria-hidden="true"
              aria-label="Profil"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="PersonIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
              />
            </svg>
          </div>
        </div>
        <div
          class="chevron"
          data-testid="toggle-sidebar"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ChevronLeftIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
          </svg>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
      data-testid="sidebar"
    >
      <div
        class="sidebar"
      >
        <div
          class="sidebar__item active"
          data-testid="side-navigate-dashboard"
        >
          <svg
            aria-hidden="true"
            aria-label="Dashbor"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-mui-internal-clone-element="true"
            data-testid="DashboardIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 13h8V3H3v10zm0 8h8v-6H3v6zm10 0h8V11h-8v10zm0-18v6h8V3h-8z"
            />
          </svg>
        </div>
        <div
          class="sidebar__item "
          data-testid="side-navigate-diary"
        >
          <svg
            aria-hidden="true"
            aria-label="Diary"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-mui-internal-clone-element="true"
            data-testid="AutoStoriesIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m19 1-5 5v11l5-4.5V1zM1 6v14.65c0 .25.25.5.5.5.1 0 .15-.05.25-.05C3.1 20.45 5.05 20 6.5 20c1.95 0 4.05.4 5.5 1.5V6c-1.45-1.1-3.55-1.5-5.5-1.5S2.45 4.9 1 6zm22 13.5V6c-.6-.45-1.25-.75-2-1v13.5c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5v2c1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.3 4.75 1.05.1.05.15.05.25.05.25 0 .5-.25.5-.5v-1.1z"
            />
          </svg>
        </div>
        <div
          class="sidebar__item "
          data-testid="side-navigate-search"
        >
          <svg
            aria-hidden="true"
            aria-label="Makanan"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-mui-internal-clone-element="true"
            data-testid="FastfoodIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18.06 22.99h1.66c.84 0 1.53-.64 1.63-1.46L23 5.05h-5V1h-1.97v4.05h-4.97l.3 2.34c1.71.47 3.31 1.32 4.27 2.26 1.44 1.42 2.43 2.89 2.43 5.29v8.05zM1 21.99V21h15.03v.99c0 .55-.45 1-1.01 1H2.01c-.56 0-1.01-.45-1.01-1zm15.03-7c0-8-15.03-8-15.03 0h15.03zM1.02 17h15v2h-15z"
            />
          </svg>
        </div>
        <div
          class="sidebar__item "
          data-testid="side-navigate-profile"
        >
          <svg
            aria-hidden="true"
            aria-label="Profil"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-mui-internal-clone-element="true"
            data-testid="PersonIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </div>
      </div>
      <div
        class="chevron"
        data-testid="toggle-sidebar"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChevronLeftIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
